app/server/lib/Authorizer.ts

```diff
@@ -193,6 +193,24 @@ export async function addRequestUser(
     }
   }
 
+  // Check if we have a boot key. This is a fallback mechanism for an
+  // administrator to authenticate themselves by demonstrating access
+  // to the environment.
+  if (!authDone && mreq.headers && mreq.headers['x-boot-key']) {
+    const reqBootKey = String(mreq.headers['x-boot-key']);
+    const bootKey = options.gristServer.getBootKey();
+    if (!bootKey || bootKey !== reqBootKey) {
```

Comment:
Do you think using a constant time comparison like this might be overkill?

https://nodejs.org/docs/latest-v18.x/api/crypto.html#cryptotimingsafeequala-b

Reply:
This is an interesting suggestion. If the boot key had a fixed length, I think I'd go for it. Given that it is variable length, it would take some time to figure out what information we've already leaked to a timing attack. There are plenty of other keys and tokens in Grist that would need to be handled similarly carefully. Using the comparison here might give an unwarranted level of confidence that this kind of attack has been considered.

Reply:
This looks like it gives full support user access to anyone with knowledge of the boot key. The boot key is visible in plaintext in URLs, right? That makes it harder to keep the key itself secret, so it does worry me a bit that someone who stumbles upon the key gets full support user access to an installation.

Reply:
Yes, the boot key is visible in URLs. The intent is for it to be used in the following cases:
 * Initial installation of Grist on a public server, prior to some form of auth being set up. This is a very vulnerable time in the life of a Grist install, and I'd like to move towards locking it down. This PR does not lock things down, but is another step on the path towards that (first step was introduction of the boot page).
 * During the lifetime of a Grist installation when there is some problem with auth.
In the admin panel, presence of a boot key will be shown as a warning or an error.

Open to other ideas for a path towards a sane way of installing Grist. I could remove the boot key for now in order to let the other steps get landed?

Reply:
Re: the second bullet, our recommendation is that the boot key be set only during such periods of maintenance, and removed promptly once problems are resolved, right? (Sort of like putting an installation in emergency or maintenance mode.)

I think that'd be ok. My main concern here would be someone leaving a key set even after initial setup is done. We could be more proactive with communication (perhaps by showing a product-wide banner whenever an installation is in "maintenance" mode). I think my main question then would be whether something like that should be done before or after landing this change. @dsagal any thoughts?